### PR TITLE
Implement support for namespaces

### PR DESCRIPTION
**Pull request by gunnarbeutner**
This PR creates a new namespace for constants (in globals.Constants). It also makes it much harder to accidentally re-define constants (even though there _are_ ways to deliberately modify them - which would be quite a bit harder to fix).

It also makes the `Types` namespace constant which makes it harder to accidentally remove or overwrite types, e.g. like this:

```
object Host "x" {
    Zone = "y" // note the incorrect spelling of the 'zone' attribute
}
```

### CODE REVIEW

**Comment by dnsmichi** [comment]
I'm expecting user config to break with this change, so when merged this must be included in the upgrading docs.

File: lib/config/expression.cpp [hunk 1]
```diff
@@ -578,6 +578,21 @@ ExpressionResult SetExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhint)
 	return Empty;
 }
 
+ExpressionResult SetConstExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhint) const
```
Comment (lazyfrosch) [inline lib/config/expression.cpp]: Please allow user to override vars like it is currently, but present a warning.

**Comment by gunnarbeutner** [comment]
Please re-test this PR.

The second version of this PR is a bit more invasive:

* New type: Namespace - which in terms of functionality is similar to the Dictionary type, except that access to individual attributes can be intercepted (via NamespaceValue and NamespaceBehavior)
* The globals dict is now an instance of the Namespace type.
* Moved constants into the System and Icinga namespaces depending on where they originally come from (libbase vs. libicinga)
* Renamed REGISTER_SCRIPTFUNCTION_*
* Made most namespaces read-only (including System, Icinga, Json, Math and Internal)
* Implemented serialization, iterator, etc. support for the Namespace type

**Comment by dnsmichi** [comment]
I like the changes a lot, and have done a feedback round already. This also solves the problems with Application::GetConst() mentioned in #6506.

File: tools/mkclass/classcompiler.cpp [hunk 2]
```diff
@@ -568,7 +568,8 @@ void ClassCompiler::HandleClass(const Klass& klass, const ClassDebugInfo&)
 		<< "{" << std::endl;
 
 	for (const Field& field : klass.Fields) {
-		m_Impl << "\t" << "Set" << field.GetFriendlyName() << "(" << "GetDefault" << field.GetFriendlyName() << "(), true);" << std::endl;
+		if (!field.PureSetAccessor)
+			m_Impl << "\t" << "Set" << field.GetFriendlyName() << "(" << "GetDefault" << field.GetFriendlyName() << "(), true);" << std::endl;
```
Comment (dnsmichi) [inline tools/mkclass/classcompiler.cpp]: I'm picking this fix into a different branch too, where I need virtual setters too.

**Comment by gunnarbeutner** [comment]
I've also now moved most configuration variables into System.Configuration:

```
<22> => System.Configuration
{
	AttachDebugger = false
	CacheDir = "/Users/gunnar/i2/var/cache/icinga2"
	Concurrency = 4.000000
	ConfigDir = "/Users/gunnar/i2/etc/icinga2"
	DataDir = "/Users/gunnar/i2/var/lib/icinga2"
	EventEngine = ""
	IncludeConfDir = "/Users/gunnar/i2/share/icinga2/include"
	InitRunDir = "/Users/gunnar/i2/var/run/icinga2"
	LocalStateDir = "/Users/gunnar/i2/var"
	LogDir = "/Users/gunnar/i2/var/log/icinga2"
	ModAttrPath = "/Users/gunnar/i2/var/lib/icinga2/modified-attributes.conf"
	ObjectsPath = "/Users/gunnar/i2/var/cache/icinga2/icinga2.debug"
	PidPath = "/Users/gunnar/i2/var/run/icinga2/icinga2.pid"
	PkgDataDir = "/Users/gunnar/i2/share/icinga2"
	PrefixDir = "/Users/gunnar/i2"
	ProgramData = ""
	RLimitFiles = 16384.000000
	RLimitProcesses = 16384.000000
	RLimitStack = 262144.000000
	RunAsGroup = "staff"
	RunAsUser = "gunnar"
	RunDir = "/Users/gunnar/i2/var/run"
	SpoolDir = "/Users/gunnar/i2/var/spool/icinga2"
	StatePath = "/Users/gunnar/i2/var/lib/icinga2/icinga2.state"
	SysconfDir = "/Users/gunnar/i2/etc"
	VarsPath = "/Users/gunnar/i2/var/cache/icinga2/icinga2.vars"
	ZonesDir = "/Users/gunnar/i2/etc/icinga2/zones.d"
	type = "Configuration"
}
```

There are a few attributes which have to stay in the global NS for now:

```
<23> => keys(globals)
[ "Environment", "Icinga", "Internal", "MaxConcurrentChecks", "NodeName", "StatsFunctions", "System", "Types" ]
```

* Environment: because users can set it with `const`
* Icinga: this is the `Icinga` NS, contains things like get_host
* Internal: internal things
* MaxConcurrentChecks: because users can set it with `const`
* NodeName: because users can set it with `const`
* StatsFunction: should probably be moved into System, oh well
* System: the `System` NS
* Types: type objects

**Comment by dnsmichi** [comment]
```
<2> => keys(globals)
[ "Environment", "Icinga", "Internal", "MaxConcurrentChecks", "NodeName", "StatsFunctions", "System", "Types" ]
<3> => keys(globals.System)
[ "ApplicationType", "ApplicationVersion", "BuildCompilerName", "BuildCompilerVersion", "BuildHostName", "Configuration", "FacilityAuth", "FacilityAuthPriv", "FacilityCron", "FacilityDaemon", "FacilityFtp", "FacilityKern", "FacilityLocal0", "FacilityLocal1", "FacilityLocal2", "FacilityLocal3", "FacilityLocal4", "FacilityLocal5", "FacilityLocal6", "FacilityLocal7", "FacilityLpr", "FacilityMail", "FacilityNews", "FacilitySyslog", "FacilityUser", "FacilityUucp", "GlobDirectory", "GlobFile", "Json", "LogCritical", "LogDebug", "LogInformation", "LogNotice", "LogWarning", "MatchAll", "MatchAny", "Math", "PlatformArchitecture", "PlatformKernel", "PlatformKernelVersion", "PlatformName", "PlatformVersion", "assert", "basename", "bool", "cidr_match", "dirname", "escape_shell_arg", "escape_shell_cmd", "exit", "get_object", "get_objects", "get_time", "glob", "glob_recursive", "intersection", "keys", "len", "log", "match", "msi_get_component_path", "number", "parse_performance_data", "path_exists", "ptr", "random", "range", "regex", "sleep", "string", "track_parents", "typeof", "union" ]
<4> => keys(globals.Icinga)
[ "Acknowledgement", "Critical", "Custom", "DbCatAcknowledgement", "DbCatCheck", "DbCatComment", "DbCatConfig", "DbCatDowntime", "DbCatEventHandler", "DbCatEverything", "DbCatExternalCommand", "DbCatFlapping", "DbCatLog", "DbCatNotification", "DbCatProgramStatus", "DbCatRetention", "DbCatState", "DbCatStateHistory", "Down", "DowntimeEnd", "DowntimeRemoved", "DowntimeStart", "FlappingEnd", "FlappingStart", "HostDown", "HostUp", "OK", "Problem", "Recovery", "ServiceCritical", "ServiceOK", "ServiceUnknown", "ServiceWarning", "Unknown", "Up", "Warning", "get_check_command", "get_event_command", "get_host", "get_host_group", "get_notification_command", "get_service", "get_service_group", "get_services", "get_time_period", "get_user", "get_user_group" ]


<6> => keys(globals.Types)
[ "ApiListener", "ApiUser", "Application", "Array", "Boolean", "CheckCommand", "CheckResult", "CheckResultReader", "Checkable", "CheckerComponent", "Command", "Comment", "CompatLogger", "ConfigObject", "Configuration", "CustomVarObject", "DateTime", "DbConnection", "Dependency", "Dictionary", "Downtime", "ElasticsearchWriter", "Endpoint", "EventCommand", "ExternalCommandListener", "FileLogger", "Function", "GelfWriter", "GraphiteWriter", "Host", "HostGroup", "IcingaApplication", "IdoMysqlConnection", "IdoPgsqlConnection", "InfluxdbWriter", "LivestatusListener", "Logger", "Namespace", "Notification", "NotificationCommand", "NotificationComponent", "Number", "Object", "OpenTsdbWriter", "PerfdataValue", "PerfdataWriter", "Reference", "ScheduledDowntime", "Service", "ServiceGroup", "StatusDataWriter", "StreamLogger", "String", "SyslogLogger", "TimePeriod", "Type", "User", "UserGroup", "Zone" ]

<7> => globals.Environment
"production"
<8> => globals.MaxConcurrentChecks
512.000000
<9> => globals.NodeName
"mbmif.int.netways.de"
```

```
<3> => globals.System.Configuration
{
	ApiBindHost = ""
	ApiBindPort = "5665"
	AttachDebugger = false
	CacheDir = "/usr/local/icinga2/var/cache/icinga2"
	Concurrency = 4.000000
	ConfigDir = "/usr/local/icinga2/etc/icinga2"
	DataDir = "/usr/local/icinga2/var/lib/icinga2"
	EventEngine = ""
	IncludeConfDir = "/usr/local/icinga2/share/icinga2/include"
	InitRunDir = "/usr/local/icinga2/var/run/icinga2"
	LocalStateDir = "/usr/local/icinga2/var"
	LogDir = "/usr/local/icinga2/var/log/icinga2"
	ModAttrPath = "/usr/local/icinga2/var/lib/icinga2/modified-attributes.conf"
	ObjectsPath = "/usr/local/icinga2/var/cache/icinga2/icinga2.debug"
	PidPath = "/usr/local/icinga2/var/run/icinga2/icinga2.pid"
	PkgDataDir = "/usr/local/icinga2/share/icinga2"
	PrefixDir = "/usr/local/icinga2"
	ProgramData = ""
	RLimitFiles = 16384.000000
	RLimitProcesses = 16384.000000
	RLimitStack = 262144.000000
	RunAsGroup = "icinga"
	RunAsUser = "icinga"
	RunDir = "/usr/local/icinga2/var/run"
	SpoolDir = "/usr/local/icinga2/var/spool/icinga2"
	StatePath = "/usr/local/icinga2/var/lib/icinga2/icinga2.state"
	SysconfDir = "/usr/local/icinga2/etc"
	VarsPath = "/usr/local/icinga2/var/cache/icinga2/icinga2.vars"
	ZonesDir = "/usr/local/icinga2/etc/icinga2/zones.d"
	type = "Configuration"
}
```

```
<4> => globals.System.Configuration.ConfigDir
"/usr/local/icinga2/etc/icinga2"
<5> => ConfigDir
"/usr/local/icinga2/etc/icinga2"
```

**Comment by dnsmichi** [comment]
The changes are transparent and not visible to the user, since the default namespaces are always included just like before. All constants and keywords are preserved and work like before.

The additional benefit of creating namespaces e.g. for custom global functions is just wonderful. This allows to better document how to work with global functions and also introduce specific includes for this. I will come up with enhanced documentation and configuration samples once this baby is merged.

```
namespace GlobalFunctions {
  function get_random_number() {
    return Math.random(10)
  }
}

using GlobalFunctions

log(get_random_number())
```

Note: This increases readability compared to `globals.get_random_number = function() { ... }`.

```
mbmif /usr/local/icinga2/etc/icinga2/tests (master *) # icinga2 daemon -C
[2018-08-24 11:37:13 +0200] information/cli: Icinga application loader (version: v2.9.1-98-g15baa8e01; debug)
[2018-08-24 11:37:13 +0200] information/cli: Loading configuration file(s).
[2018-08-24 11:37:14 +0200] information/config: 0.000008
[2018-08-24 11:37:14 +0200] information/ConfigItem: Committing config item(s).
```

**Pull request merged**